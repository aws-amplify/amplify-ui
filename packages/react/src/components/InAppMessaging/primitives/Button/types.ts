// import { PressableProps, StyleProp, TextStyle } from 'react-native';

export interface ButtonProps {
  style?: unknown;
  textStyle?: unknown;
}

export interface ButtonStyles {
  text: unknown;
}
