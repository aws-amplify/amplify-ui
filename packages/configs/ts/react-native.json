{
  "extends": "./base.json",
  "compilerOptions": {
    "jsx": "react-native",
    "skipLibCheck": true,
    "target": "esnext"
  }
}
