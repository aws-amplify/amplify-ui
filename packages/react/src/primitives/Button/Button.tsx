import classNames from 'classnames';
import * as React from 'react';

import { classNameModifier, classNameModifierByFlag } from '../shared/utils';
import {
  BaseButtonProps,
  ButtonProps,
  ForwardRefPrimitive,
  Primitive,
} from '../types';
import { ComponentClassNames } from '../shared/constants';
import { Flex } from '../Flex';
import { Loader } from '../Loader';
import { View } from '../View';

const ButtonPrimitive: Primitive<ButtonProps, 'button'> = (
  {
    className,
    children,
    isFullWidth = false,
    isDisabled,
    isLoading,
    loadingText = '',
    size,
    type = 'button',
    variation,
    ...rest
  },
  ref
) => {
  const componentClasses = classNames(
    ComponentClassNames.Button,
    ComponentClassNames.FieldGroupControl,
    classNameModifier(ComponentClassNames.Button, variation),
    classNameModifier(ComponentClassNames.Button, size),
    classNameModifierByFlag(
      ComponentClassNames.Button,
      'disabled',
      isDisabled ?? isLoading ?? rest['disabled']
    ),
    classNameModifierByFlag(ComponentClassNames.Button, 'loading', isLoading),
    classNameModifierByFlag(
      ComponentClassNames.Button,
      'fullwidth',
      isFullWidth
    ),
    className
  );

  return (
    <View
      ref={ref}
      as="button"
      className={componentClasses}
      data-fullwidth={isFullWidth}
      data-loading={isLoading}
      data-size={size}
      data-variation={variation}
      isDisabled={isDisabled ?? isLoading}
      type={type}
      {...rest}
    >
      {isLoading ? (
        <Flex as="span" className={ComponentClassNames.ButtonLoaderWrapper}>
          <Loader size={size} />
          {loadingText ? loadingText : null}
        </Flex>
      ) : (
        children
      )}
    </View>
  );
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/button)
 */
export const Button: ForwardRefPrimitive<BaseButtonProps, 'button'> =
  React.forwardRef(ButtonPrimitive);

Button.displayName = 'Button';
