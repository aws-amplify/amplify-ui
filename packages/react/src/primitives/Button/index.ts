export * from './Button'
export * from './demo';
