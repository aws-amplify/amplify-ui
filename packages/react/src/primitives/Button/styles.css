/**
  * Amplify Button styles
  * @TODO: Generate this file from Style dictionary
*/
@import "@aws-amplify/ui-core/theme.css";

.amplify-ui-button {
  /* Button */
  --amplify-ui-button-padding: var(--amplify-ui-sizes-4);
  --amplify-ui-button-disabled-opacity: 0.4;
  --amplify-ui-button-disabled-cursor: not-allowed;
  --amplify-ui-button-transition: all 250ms;
  --amplify-ui-button-font-weight: var(--amplify-ui-fontWeights-semibold);

  --amplify-ui-button-colors-text-primary: var(--amplify-ui-colors-white);
  --amplify-ui-button-colors-text-tertiary: var(--amplify-ui-colors-black);
  --amplify-ui-button-colors-bg-primary: var(--amplify-ui-colors-purple-500);
  --amplify-ui-button-colors-border-primary: var(
    --amplify-ui-colors-purple-500
  );
  --amplify-ui-button-colors-bg-primary-hover: var(
    --amplify-ui-colors-purple-400
  );
  --amplify-ui-button-colors-bg-primary-loading: var(
    --amplify-ui-colors-purple-400
  );
  --amplify-ui-button-height-primary: var(--amplify-ui-sizes-12);

  --amplify-ui-button-colors-text-secondary: #000;
  --amplify-ui-button-colors-border-secondary: var(
    --amplify-ui-colors-purple-200
  );
  --amplify-ui-button-border-secondary: 1px solid
    var(--amplify-ui-button-colors-border-secondary);
  --amplify-ui-button-colors-bg-secondary-hover: var(
    --amplify-ui-colors-purple-50
  );

  --amplify-ui-button-colors-text-tertiary: var(--amplify-ui-colors-black);
  --amplify-ui-button-colors-bg-tertiary: transparent;
  --amplify-ui-button-colors-bg-tertiary-hover: var(
    --amplify-ui-colors-purple-200
  );

  --amplify-ui-button-colors-text-tertiary: var(--amplify-ui-colors-black);
  --amplify-ui-button-colors-text-link: var(--amplify-ui-colors-purple-500);

  --amplify-ui-button-height-large: var(--amplify-ui-sizes-12);
  --amplify-ui-button-height-medium: var(--amplify-ui-sizes-10);
  --amplify-ui-button-height-small: var(--amplify-ui-sizes-8);

  --amplify-ui-button-font-size-large: var(--amplify-ui-fontSizes-lg);
  --amplify-ui-button-font-size-medium: var(--amplify-ui-fontSizes-md);
  --amplify-ui-button-font-size-small: var(--amplify-ui-fontSizes-sm);

  all: unset; /* protect against external styles */
  font-weight: var(--amplify-ui-button-font-weight);
  color: var(--amplify-ui-colors-black);
  transition: var(--amplify-ui-button-transition);
  border-radius: var(--amplify-ui-border-radius-md);
  box-sizing: border-box;
}
.amplify-ui-button:hover {
  cursor: pointer;
}
.amplify-ui-button[disabled] {
  cursor: var(--amplify-ui-button-disabled-cursor);
  opacity: var(--amplify-ui-button-disabled-opacity);
}
.amplify-ui-button[data-fullwidth="true"] {
  width: 100%;
}
.amplify-ui-button[data-variant="primary"] {
  color: var(--amplify-ui-button-colors-text-primary);
  background-color: var(--amplify-ui-button-colors-bg-primary);
}
.amplify-ui-button[data-variant="primary"][data-loading="true"] {
  background-color: var(--amplify-ui-button-colors-bg-primary-loading);
}
.amplify-ui-button[data-variant="primary"]:hover {
  background-color: var(--amplify-ui-button-colors-bg-primary-hover);
}
.amplify-ui-button[data-variant="primary"]:hover[disabled] {
  background-color: var(--amplify-ui-button-colors-bg-primary);
}
.amplify-ui-button[data-variant="secondary"] {
  color: var(--amplify-ui-button-colors-text-secondary);
  border: var(--amplify-ui-button-border-secondary);
}
.amplify-ui-button[data-variant="secondary"]:hover {
  background-color: var(--amplify-ui-button-colors-bg-secondary-hover);
}
.amplify-ui-button[data-variant="tertiary"]:hover,
.amplify-ui-button[data-variant="tertiary"][disabled] {
  background-color: var(--amplify-ui-button-colors-bg-tertiary-hover);
}
.amplify-ui-button[data-variant="link"] {
  color: var(--amplify-ui-button-colors-text-link);
}
.amplify-ui-button[data-variant="link"]:hover {
  text-decoration: underline;
}
.amplify-ui-button[data-size="small"] {
  padding: 0 var(--amplify-ui-sizes-3);
  height: var(--amplify-ui-button-height-small);
  font-size: var(--amplify-ui-button-font-size-small);
  border-radius: var(--amplify-ui-border-radius-sm);
}
.amplify-ui-button[data-size="medium"] {
  padding: 0 var(--amplify-ui-sizes-4);
  height: var(--amplify-ui-button-height-medium);
  font-size: var(--amplify-ui-button-font-size-medium);
  border-radius: var(--amplify-ui-border-radius-md);
}
.amplify-ui-button[data-size="large"] {
  padding: 0 var(--amplify-ui-sizes-5);
  height: var(--amplify-ui-button-height-large);
  font-size: var(--amplify-ui-button-font-size-large);
  border-radius: var(--amplify-ui-border-radius-lg);
}
