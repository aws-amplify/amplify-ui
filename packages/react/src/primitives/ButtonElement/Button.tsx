import React from "react";
import { ButtonProps, ButtonSize, ButtonVariant } from "../shared/types";
export const ButtonElement: React.FC<ButtonProps> = props => {
  const {
    active,
    ariaLabel,
    children,
    disabled,
    id,
    loading,
    loadingText,
    size = ButtonSize.Medium,
    variant = ButtonVariant.Secondary,
    fullWidth = false,
    onClick = () => {}
  } = props;
  return (
    <button
      id={id}
      className={`amplify-ui-button ${props.className}`}
      disabled={disabled || loading}
      data-loading={loading}
      data-fullwidth={fullWidth}
      data-active={active}
      data-size={size}
      data-variant={variant}
      onClick={onClick}
      aria-label={ariaLabel}
    >
      {loading ? <span>{loadingText}</span> : children}
    </button>
  );
};
