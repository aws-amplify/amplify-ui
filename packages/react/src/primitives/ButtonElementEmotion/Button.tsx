import React from "react";
import styled, { CSSObject } from "@emotion/styled";
import { useTheme } from "@emotion/react";
import { ButtonProps, ButtonSize, ButtonVariant } from "../shared/types";
import { AmplifyComponent } from "../AmplifyUIProvider/AmplifyUI";

const getEmotionStateAttributes = (styles: CSSObject) => {
  let current = styles;
  const keys = Object.keys(current);
  for (let key of keys) {
    if (key === "_hover") {
      current["&:hover"] = current["_hover"];
    }
    if (key === "_disabled") {
      current["&[disabled]"] = current["_disabled"];
    }
  }
  return styles;
};

const getButtonStyles = (theme, props: ButtonProps): CSSObject => {
  // @TODO: still needs some work to merge nested objects
  // (baseStyles and variant _disabled styles, for example)

  const { baseStyle, variants, sizes, states } = theme.components.Button;
  let styles = { ...baseStyle };
  if (props.variant in variants) {
    styles = { ...styles, ...variants[props.variant] };
  }
  if (props.size in sizes) {
    styles = { ...styles, ...sizes[props.size] };
  }
  if (props.loading && states?.loading != null) {
    styles = { ...styles, ...states.loading };
  }
  return getEmotionStateAttributes(styles);
};

export const ButtonElementEmotion: React.FC<ButtonProps> = props => {
  const {
    active,
    ariaLabel,
    children,
    disabled,
    id,
    loading,
    loadingText = "Loading...",
    size = ButtonSize.Medium,
    variant = ButtonVariant.Secondary,
    fullWidth = false,
    onClick = () => {}
  } = props;
  const theme = useTheme();
  const styles = getButtonStyles(theme, { variant, size, loading });

  const Button = styled.button(() => {
    return styles;
  });

  return (
    <AmplifyComponent>
      <Button
        id={id}
        className={props.className}
        disabled={disabled || loading}
        data-loading={loading}
        data-fullwidth={fullWidth}
        data-active={active}
        data-size={size}
        data-variant={variant}
        onClick={onClick}
        aria-label={ariaLabel}
      >
        {loading ? <span>{loadingText}</span> : children}
      </Button>
    </AmplifyComponent>
  );
};
