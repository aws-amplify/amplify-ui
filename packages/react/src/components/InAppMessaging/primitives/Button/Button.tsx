import React from 'react';
import { Pressable, Text } from 'react-native';

import { styles } from './styles';
import { ButtonProps } from './types';

export default function Button({
  children,
  textStyle,
  ...pressableProps
}: ButtonProps): JSX.Element {
  return (
    <Pressable {...pressableProps}>
      {typeof children === 'string' ? (
        <Text style={[styles.text, textStyle]}>{children}</Text>
      ) : (
        children
      )}
    </Pressable>
  );
}
