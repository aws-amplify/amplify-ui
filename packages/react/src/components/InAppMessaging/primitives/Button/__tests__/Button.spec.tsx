import React from 'react';
import { Text } from 'react-native';
import TestRenderer from 'react-test-renderer';

import Button from '../Button';

const title = 'A preesable button';
const Title = () => <Text>{title}</Text>;

describe('Button', () => {
  it('renders as expected with a string passed as children', () => {
    const button = TestRenderer.create(<Button>{title}</Button>);
    expect(button.toJSON()).toMatchSnapshot();
  });

  it('renders as expected with a component passed as children', () => {
    const button = TestRenderer.create(
      <Button>
        <Title />
      </Button>
    );
    expect(button.toJSON()).toMatchSnapshot();
  });
});
