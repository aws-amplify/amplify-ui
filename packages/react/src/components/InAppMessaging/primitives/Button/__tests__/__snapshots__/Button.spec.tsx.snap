// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders as expected with a component passed as children 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text>
    A preesable button
  </Text>
</View>
`;

exports[`Button renders as expected with a string passed as children 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
        },
        undefined,
      ]
    }
  >
    A preesable button
  </Text>
</View>
`;
